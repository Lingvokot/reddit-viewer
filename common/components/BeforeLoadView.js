import React, { Component } from "react";
import { Button, View, Text } from "react-native";

export default class BeforeLoadView extends Component {
	render() {
		if (!this.props.finished)
			return (
				<Text>Loading posts, please wait...</Text>
			);
		else if (this.props.error)
			return (
				<View>
					<Text>{this.props.error}</Text>
					<Button title="Retry to load"
							onPress={this.props.retryFunc}
							style={{
								paddingTop: 5, paddingBottom: 5,
								margin: 5
							}} ref={(ref) => {this.retryButton = ref;}}/>
				</View>
			);
		else
			return null;
	}
}