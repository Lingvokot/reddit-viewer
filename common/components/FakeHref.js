import React, { Component, PropTypes } from "react";
import { TouchableHighlight, Linking, Text } from "react-native";

export default class FakeHref extends Component {
	render() {
		return (
			<TouchableHighlight
				onPress={() => {
					return Linking.canOpenURL(this.props.url).then(supported => {
						if (!supported)
							this.props.onCannotHandle();
						else
							return Linking.openURL(this.props.url);
					}).catch(this.props.errorHandler);
				}} ref={(ref) => {this.highlight = ref;}}>
				<Text style={{color: "blue", textDecorationLine: "underline"}}>
					{this.props.url}
				</Text>
			</TouchableHighlight>
		);
	}
}
