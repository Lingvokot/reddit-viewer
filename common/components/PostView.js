/*global alert*/
import React, { Component, PropTypes } from "react";
import { View, Text, Button, Modal } from "react-native";
import FakeHref from "./FakeHref";

export default class PostView extends Component {
	constructor(props) {
		super(props);
		this.state = {urlOpenError: null};
	}
	render() {
		let selectedPost = this.props.selectedPost;
		return (
			<View>
				<Button onPress={this.props.onBack}
						title="Back"
						style={{
							paddingTop: 5, paddingBottom: 5,
							backgroundColor: "blue"
						}} ref={(ref) => {this.backButton = ref;}}/>
				<Text style={{fontWeight: "bold", textAlign: "center"}}>
					{selectedPost.title}
				</Text>
				<Text>{`ID: ${selectedPost.name}`}</Text>
				<Text>{`Author: ${selectedPost.author}`}</Text>
				<Text>{`Score: ${selectedPost.score}`}</Text>
				<Text>External URL:</Text>
				<FakeHref url={selectedPost.cleanUrl}
					onCannotHandle={() => {
						alert("Cannot handle this URL");
					}}
					errorHandler={(err) => {
						alert(err.message);
					}} ref={(ref) => {this.externalLink = ref;}}/>
			</View>
		);
	}
}