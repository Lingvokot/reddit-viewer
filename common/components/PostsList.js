import React, { Component, PropTypes } from 'react';
import { View, Text, TouchableHighlight, ListView } from 'react-native';

export default class PostsList extends Component {
	render() {
		this.listElements = [];
		if (this.props.toList.getRowCount() == 0)
			return (
				<Text>There are no posts yet.</Text>
			)
		return (
			<ListView dataSource={this.props.toList}
				onEndReached={() => {
					this.props.onEndReached();
				}}
				renderRow={(rowData) => (
					<TouchableHighlight style={{
							borderWidth: 1,
							borderRadius: 3,
							borderColor: "red",
							margin: 5,
							padding: 5
						}} ref={(ref) => {this.listElements.push(ref);}}
						onPress={() => {this.props.onForward(rowData);}}>
						<Text>{rowData.title}</Text>
					</TouchableHighlight>
				)} ref={(ref) => {this.listView = ref;}}/>
		);
	}
}