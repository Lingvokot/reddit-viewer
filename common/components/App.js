import React, { Component } from 'react';
import { Navigator, ListView } from 'react-native';
import PostsList from './PostsList';
import PostView from './PostView';
import BeforeLoadView from './BeforeLoadView';
import { last } from 'lodash/array';
import { each } from 'lodash/collection';
import loadPosts from './loadPosts';

const cardsInPage = 10;

export default class App extends Component {
	constructor(props) {
		super(props);
		this.state = {
			selectedPost: null, postsToList: [], pagesFetched: 0,
			error: null, finished: false, inProcess: false
		};
	}
	componentWillMount() {
		if (this.state.postsToList.length == 0)
			this.loadMorePosts();
	}
	loadMorePosts() {
		if (this.state.inProcess)
			return;
		return new Promise((resolve) => {
			this.setState({finished: false, inProcess: true}, async () => {
				let page = null;
				try {
					if (this.state.postsToList.length == 0)
						page = await loadPosts({limit: 10});
					else {
						const after = last(this.state.postsToList).uuid;
						page = await loadPosts({after, limit: 10});
					}
				}
				catch (e) {
					this.setState({error: e.message, finished: true, inProcess: false}, () => resolve());
					return;
				}
				let newPosts = Object.values(page.apiResponse.posts);
				this.setState({
					error: null, finished: true, inProcess: false,
					postsToList: this.state.postsToList.concat(newPosts),
					pagesFetched: this.state.pagesFetched + 1
				}, () => resolve());
			});
		});
	}
	render() {
		if (this.state.error || (!this.state.finished && this.state.postsToList.length == 0))
			return (
				<BeforeLoadView error={this.state.error}
					finished={this.state.finished}
					retryFunc={() => this.loadMorePosts()}/>
			)
		return (
			<Navigator initialRoute={{ title: 'Posts', index: 0 }}
				renderScene={(route, navigator) => {
				const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
				if (route.index == 0)
					return (
						<PostsList toList={ds.cloneWithRows(this.state.postsToList)}
							onForward={(selectedPost) => {
								return new Promise(resolve => {
									this.setState({selectedPost});
									resolve();
									navigator.push({title: selectedPost.title, index: route.index + 1});
								}).catch(err => {});
							}}
							onEndReached={() => this.loadMorePosts()}
							ref={(ref) => {this.postsList = ref;}}/>
					)
				else
					return (
						<PostView selectedPost={this.state.selectedPost}
							onBack={() => {
								return new Promise(resolve => {
									this.setState({selectedPost: null});
									resolve();
									navigator.pop();
								}).catch(err => {});
							}} ref={(ref) => {this.postView = ref;}}/>
					)
			}} />
		);
	}
}