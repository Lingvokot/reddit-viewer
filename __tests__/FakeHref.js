/*global jest, describe, it, test, jasmine, expect, console*/
/*eslint no-console: "off"*/
import "react-native";
import {ListView, Linking} from "react-native";
import React from "react";
import Index from "../index.android.js";
import App from "../common/components/App";
import BeforeLoadView from "../common/components/BeforeLoadView";
import PostView from "../common/components/PostView";
import PostsList from "../common/components/PostsList";
import FakeHref from "../common/components/FakeHref";
const {configureMock, makeArtificialErrorEvent} = Linking;

// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";
jest.useFakeTimers();
jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;

describe("Testing FakeHref", () => {
	const cleanUrl = "http://www.businessinsider.com.au/elijah-wood-hollywood-pedophile-ring-2016-5?r=US&IR=T";

	test("General rendering", () => {
		const mockOnCannotHandle = jest.fn();
		mockOnCannotHandle.mockImplementation(() => {});
		const mockOnError = jest.fn();
		mockOnError.mockImplementation((error) => {
			console.log(`Error "${error.message}" is successfully caught`);
		});
		const domTree = renderer.create(
			<FakeHref url={cleanUrl}
				onCannotHandle={mockOnCannotHandle}
				errorHandler={mockOnError}/>
		);
		const highlight = domTree.getInstance().highlight;
		const tree = domTree.toJSON();
		expect(tree).toMatchSnapshot();
		expect(typeof highlight.props.onPress).toEqual("function");
	});
	test("Touch handling: some internal will occur if we try to open URL, canOpenURL returns false", (done) => {
		configureMock(false, true);
		const mockOnCannotHandle = jest.fn();
		mockOnCannotHandle.mockImplementation(() => {
			makeArtificialErrorEvent({error: "Cannot handle"});
		});
		const mockOnError = jest.fn();
		mockOnError.mockImplementation((error) => {
			console.log(`Error "${error.message}" is successfully caught`);
		});
		const domTree = renderer.create(
			<FakeHref url={cleanUrl}
				onCannotHandle={mockOnCannotHandle}
				errorHandler={mockOnError}/>
		);
		const highlight = domTree.getInstance().highlight;
		Linking.addEventListener("url", (event) => {
			expect(event.error).toEqual("Cannot handle");
			expect(mockOnError).not.toBeCalled();
			Linking.removeAllEventListeners();
			done();
		});
		highlight.props.onPress();
	});
	test("Touch handling: some internal error will occur if we open URL but canOpenURL returns true", (done) => {
		configureMock(true, true);
		const mockOnCannotHandle = jest.fn();
		mockOnCannotHandle.mockImplementation(() => {expect(true).toBeFalsy();});
		const mockOnError = jest.fn();
		mockOnError.mockImplementation((error) => {
			console.log(`Error "${error.message}" is successfully caught`);
			done();
		});
		const domTree = renderer.create(
			<FakeHref url={cleanUrl}
				onCannotHandle={mockOnCannotHandle}
				errorHandler={mockOnError}/>
		);
		const highlight = domTree.getInstance().highlight;
		highlight.props.onPress();
	});
	test("Touch handling: all must be OK", (done) => {
		configureMock(true, false);
		const mockOnCannotHandle = jest.fn();
		mockOnCannotHandle.mockImplementation(() => {});
		const mockOnError = jest.fn();
		mockOnError.mockImplementation((error) => {
			console.log(`Error "${error.message}" is successfully caught`);
		});
		const domTree = renderer.create(
			<FakeHref url={cleanUrl}
				onCannotHandle={mockOnCannotHandle}
				errorHandler={mockOnError}/>
		);
		const highlight = domTree.getInstance().highlight;
		Linking.addEventListener("url", (event) => {
			expect(event.error).toBeUndefined();
			expect(event.url).toEqual(cleanUrl);
			expect(mockOnCannotHandle).not.toBeCalled();
			expect(mockOnError).not.toBeCalled();
			Linking.removeAllEventListeners();
			done();
		});
		highlight.props.onPress();
	});
});
