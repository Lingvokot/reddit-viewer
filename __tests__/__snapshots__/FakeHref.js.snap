// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing FakeHref General rendering 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      undefined,
    ]
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "blue",
        "textDecorationLine": "underline",
      }
    }
  >
    http://www.businessinsider.com.au/elijah-wood-hollywood-pedophile-ring-2016-5?r=US&IR=T
  </Text>
</View>
`;
